## Serving predictions on a web interface

First, Now we will buid the model with reduced_df for web app creation

In [1]:
import pandas as pd

reduced_df = pd.read_csv('reduced_df.csv')

reduced_df

Unnamed: 0     Store  DayOfWeek        Date     Sales  Customers  \
0           947227 -0.917744  -1.501129  2013-03-04 -0.473833  -0.653996   
1           490650  0.862265   0.501484  2014-04-18 -1.582625  -1.572631   
2           801072 -1.197327   1.002138  2013-07-13 -0.051953   0.274932   
3           154587  0.495701   1.502791  2015-03-15 -1.582625  -1.572631   
4           125375 -0.193936   0.501484  2015-04-10  0.249510   0.298091   
...            ...       ...        ...         ...       ...        ...   
508599      804881  0.244076  -0.499823  2013-07-10  0.330159   0.051063   
508600      610981  0.585788  -1.000476  2013-12-31 -0.432674  -0.489311   
508601      833094  1.294064   1.002138  2013-06-15 -0.279440  -0.208831   
508602      535773 -0.976767   1.002138  2014-03-08 -0.320877  -0.103330   
508603      717165  1.390365   0.501484  2013-09-27 -0.022752   0.532253   

        Open  Promo StateHoliday  SchoolHoliday  ... Promo_3 Promo_4  Promo_5  \
0          1      1            0              0  ...       0       1        0   
1          0      1            b              1  ...       0       0        0   
2          1      0            0              0  ...       0       0        0   
3          0      0            0              0  ...       0       0        0   
4          1      0            0              1  ...       0       0        0   
...      ...    ...          ...            ...  ...     ...     ...      ...   
508599     1      0            0              0  ...       0       0        1   
508600     1      0            0              1  ...       0       0        0   
508601     1      0            0              0  ...       0       1        0   
508602     1      0            0              0  ...       0       1        0   
508603     1      1            0              0  ...       0       0        1   

        Promo_6  Promo_7  Promo_8  Promo_9  Promo_10 Promo_11  Promo_12  
0             0        1        0        0         1        0         0  
1             0        0        0        0         0        0         0  
2             0        0        0        0         0        0         0  
3             0        0        0        0         0        0         0  
4             0        0        0        0         0        0         0  
...         ...      ...      ...      ...       ...      ...       ...  
508599        0        0        1        0         0        1         0  
508600        0        0        0        0         0        0         0  
508601        0        1        0        0         1        0         0  
508602        0        1        0        0         1        0         0  
508603        0        0        1        0         0        1         0  

[508604 rows x 36 columns]

In [2]:
# Now we will buid the model with reduced_df for web app use

# Again now Let's start by defining our features and target, and then splitting the data into training and testing sets

from sklearn.model_selection import train_test_split

# Defining features and target
X = reduced_df.drop(['Sales', 'Date', 'StoreType', 'Assortment', 'PromoInterval', 'StateHoliday'], axis=1)
y = reduced_df['Sales']

# Splitting the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

X_train.shape, X_test.shape, y_train.shape, y_test.shape

((406883, 30), (101721, 30), (406883,), (101721,))

The dataset has been split into training and testing sets, with 406883 samples for training and 101721 samples for testing.

In [3]:
# Now again, let's create a pipeline that includes the Random Forest Regressor. 
# We'll train the model using the training data and then evaluate its performance on the test set. 

from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestRegressor

# Creating the pipeline with a Random Forest Regressor
pipeline = Pipeline([
    ('rf', RandomForestRegressor(n_estimators=100, random_state=42))
])

# Training the model
pipeline.fit(X_train, y_train)

# Predicting on the test set
y_pred = pipeline.predict(X_test)


#2.3 Choosing a loss function (Mean Squared Error (MSE))
from sklearn.metrics import mean_squared_error, r2_score

# Evaluating the model
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

mse, r2

(0.01430401413391984, 0.9857552878290172)

MSE close to 0 and an R-squared value close to 1 together indicate that the model is performing very well, with high predictive accuracy and a strong fit to the data.

Serialize the model

In [4]:
import joblib
from datetime import datetime


model = pipeline  

# Get the current time and format it as MM-DD-YYYY-HH-MM-SS-00
timestamp = datetime.now().strftime("%m-%d-%Y-%H-%M-%S-00")

# Construct the filename with the timestamp
filename = f"Random_forest_model_{timestamp}.pkl"

# Save the model
with open(filename, 'wb') as file:
    joblib.dump(model, file)

print(f"Model saved as {filename}")

Model saved as Random_forest_model_08-19-2024-00-09-27-00.pkl


Load the model

In [6]:
import joblib

# Load the model from the file
with open('Random_forest_model_08-19-2024-00-09-27-00.pkl', 'rb') as file:
    loaded_model = joblib.load(file)

# loaded_model is now ready to use

# Check if the model is loaded correctly
print(type(loaded_model))

<class 'sklearn.pipeline.Pipeline'>


In [7]:
loaded_model

Pipeline(steps=[('rf', RandomForestRegressor(random_state=42))])